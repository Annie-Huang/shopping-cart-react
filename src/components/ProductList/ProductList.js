import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import * as shoppingCartActions from '../../actions/shoppingCartActions';

export class ProductList extends Component {
    updateItemInCart = data => {
        this.props.updateItemInCart(data);
    };

    render() {
        const children = this.props.products.map(product =>
            <div className="card" key={product.id}>
                <div className="card-header text-center text-white bg-primary">
                    <h4 className="card-title">{product.name}</h4>
                </div>
                <div className="card-body">
                    <div className='row'>
                        <div className='col-md-3'>Price:</div>
                        <div className='col-md-7'>${product.price}</div>
                    </div>
                </div>
                <div className="card-footer">
                    <button type="button"
                            className="btn btn-primary"
                            onClick={()=>this.updateItemInCart({product, quantity: 1})}
                    >
                        Add 1 item
                    </button>
                    &nbsp;&nbsp;
                    {product.productInCart &&
                        <button type="button"
                                className="btn btn-secondary"
                                onClick={() => this.updateItemInCart({product, quantity: -1})}
                        >
                            Remove 1 item
                        </button>
                    }
                </div>
            </div>
        );

        return (
            this.props.showProductList &&
            <React.Fragment>
                <div className="card-deck">
                    {children}
                </div>
                <br/>
            </React.Fragment>
        )
    }
}

ProductList.propTypes = {
    products: PropTypes.array.isRequired,
    showProductList: PropTypes.bool.isRequired,
    updateItemInCart: PropTypes.func.isRequired
};

export const updateProductsWithInCartInfo = (products, cartItems) => {
    const updatedProducts = [];
    products.forEach(product => {
        const matchCartItem = cartItems.find((cartItem) => cartItem.product.id === product.id);
        updatedProducts.push({...product, productInCart: !!matchCartItem});
    });

    return updatedProducts;
};

const mapStateToProps = (state) => ({
    products: updateProductsWithInCartInfo(state.products, state.cartItems),
    showProductList: !!state.user.id,
});

const mapDispatchToProps = (dispatch) => ({
    updateItemInCart: data => dispatch(shoppingCartActions.updateItemInCart(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ProductList);
