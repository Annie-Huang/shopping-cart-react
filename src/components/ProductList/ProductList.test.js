import React from 'react';
import {shallow, mount} from 'enzyme';
import {ProductList, updateProductsWithInCartInfo} from './ProductList';

describe('ProductList presentation', () => {
    const updateItemInCart = jest.fn();
    const products = require('../../resources/fixtures/products.json');
    products[0].productInCart = true;
    products[1].productInCart = false;
    products[2].productInCart = true;
    const showProductList = true;
    const props = {
        updateItemInCart,
        products,
        showProductList
    };

    it('should render ProductList component', () => {
        expect(shallow(<ProductList {...props} />)).toMatchSnapshot();
    });

    it('#updateItemInCart -- called directly', () => {
        const wrapper = shallow(<ProductList {...props} />).instance();
        const data = {
            product: products[0],
            quantity: 1
        };
        wrapper.updateItemInCart(data);

        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateItemInCart should be called when clicking Add 1 item button', () => {
        const wrapper = shallow(<ProductList {...props} />);

        const buttons = wrapper.find('button');
        const addButton =  buttons.first();
        expect(addButton.text()).toBe('Add 1 item');

        addButton.simulate('click');
        const data = {
            product: products[0],
            quantity: 1
        };
        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateItemInCart should be called when clicking Remove item button', () => {
        const wrapper = shallow(<ProductList {...props} />);

        const buttons = wrapper.find('button');
        const removeButton =  buttons.at(1);
        expect(removeButton.text()).toBe('Remove 1 item');
        removeButton.simulate('click');

        const data = {
            product: products[0],
            quantity: -1
        };
        expect(updateItemInCart).toBeCalledWith(data);
    });

    it('#updateProductsWithInCartInfo should be productInCart property into each product', () => {
        const originalProducts = require('../../resources/fixtures/products.json');
        const cartItems = require('../../resources/fixtures/cart-items.json');

        const updatedProducts = updateProductsWithInCartInfo(originalProducts, cartItems);
        expect(updatedProducts[0].productInCart).toBe(true);
        expect(updatedProducts[1].productInCart).toBe(true);
        expect(updatedProducts[2].productInCart).toBe(false);
    });

});
