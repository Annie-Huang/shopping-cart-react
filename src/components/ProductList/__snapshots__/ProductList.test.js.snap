// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList presentation should render ProductList component 1`] = `
<Fragment>
  <div
    className="card-deck"
  >
    <div
      className="card"
      key="classic"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Classic Ad
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            269.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Remove 1 item
        </button>
      </div>
    </div>
    <div
      className="card"
      key="standout"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Standout Ad
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            322.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
      </div>
    </div>
    <div
      className="card"
      key="premium"
    >
      <div
        className="card-header text-center text-white bg-primary"
      >
        <h4
          className="card-title"
        >
          Premium Ad
        </h4>
      </div>
      <div
        className="card-body"
      >
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            Price:
          </div>
          <div
            className="col-md-7"
          >
            $
            394.99
          </div>
        </div>
      </div>
      <div
        className="card-footer"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          Add 1 item
        </button>
          
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Remove 1 item
        </button>
      </div>
    </div>
  </div>
  <br />
</Fragment>
`;
